### PR TITLE
Add kernel (mass) with non-constant coefficients

### PR DESCRIPTION
**Pull request by SarahRo**
Implements (c_1(x)+c_2(x)+ ...)*u*v*dx.

### CODE REVIEW

File: cpp/kernels_non_const_coefficient.hpp [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+  // Get coordinate element from dolfinx
+  basix::FiniteElement coordinate_element = basix::create_element("Lagrange", cell, 1);
+  xt::xtensor<double, 4> coordinate_basis = coordinate_element.tabulate(1, points);
+
+  xt::xtensor<double, 2> dphi0_c
+      = xt::view(coordinate_basis, xt::range(1, tdim + 1), 0, xt::all(), 0);
+
+  assert(ndofs_cell == static_cast<std::int32_t>(phi.shape(1)));
+
+  // Stiffness Matrix using quadrature formulation
```
Comment (jorgensd) [inline cpp/kernels_non_const_coefficient.hpp:88]: Remove commented code

File: cpp/kernels_non_const_coefficient.hpp [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+      for (int i = 0; i < ndofs_cell; i++)
+      {
+        double w1 = w0 * phi.unchecked(q, i);
+        for (int j = 0; j < ndofs_cell; j++)
+          A[i * ndofs_cell + j] += w1 * phi.unchecked(q, j);
+      }
+    }
+  };
+
+  // // Mass Matrix using tensor contraction formulation
```
Comment (jorgensd) [inline cpp/kernels_non_const_coefficient.hpp:177]: Same as above

File: cpp/kernels_non_const_coefficient.hpp [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+  //       }
+  //     }
+  //   }
+  // };
+
+  switch (type)
+  {
+  case dolfinx_cuas::Kernel::Mass:
+    return mass_coeff;
+  // case dolfinx_cuas::Kernel::MassTensor:
```
Comment (jorgensd) [inline cpp/kernels_non_const_coefficient.hpp:330]: Same as above

File: cpp/kernels_non_const_coefficient.hpp [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+} // namespace
+namespace dolfinx_cuas
+{
+
+/// Create integration kernel for Pth order Lagrange elements
+/// @param[in] type The kernel type (Mass or Stiffness)
+/// @param[in] P Degree of the element
+/// @param[in] bs The block size
+/// @return The integration kernel
+kernel_fn generate_coeff_kernel(
```
Comment (jorgensd) [inline cpp/kernels_non_const_coefficient.hpp:351]: Have you tested this with block size?
The kernels Igor implemented does not support it. I have a branch where this is about to change

Comment (SarahRo) [reply cpp/kernels_non_const_coefficient.hpp:141]: No. This is based on Igor's kernel. I can wait with merging and add your changes.

File: cpp/kernels_non_const_coefficient.hpp [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+  // NOTE: These assumptions are only fine for simplices
+  int quad_degree = 0;
+  if (type == dolfinx_cuas::Kernel::Stiffness)
+    quad_degree = (P - 1) + (P - 1);
+  else if (type == dolfinx_cuas::Kernel::Mass or type == dolfinx_cuas::Kernel::MassTensor)
+    quad_degree = 2 * P + Q;
+  else if (type == dolfinx_cuas::Kernel::TrEps)
+    quad_degree = (P - 1) + (P - 1);
+  else if (type == dolfinx_cuas::Kernel::SymGrad)
+    quad_degree = (P - 1) + (P - 1);
```
Comment (jorgensd) [inline cpp/kernels_non_const_coefficient.hpp:48]:
```suggestion  
   const int quad_degree = 2 * P + Q;
```

File: cpp/kernels_non_const_coefficient.hpp [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+
+  auto [points, weight]
+      = basix::quadrature::make_quadrature("default", basix::cell::str_to_type(cell), quad_degree);
+  std::vector<double> weights(weight);
+
+  // Create Finite element for test and trial functions and tabulate shape functions
+  basix::FiniteElement element = basix::create_element(family, cell, P);
+  xt::xtensor<double, 4> basis = element.tabulate(1, points);
+  xt::xtensor<double, 2> phi = xt::view(basis, 0, xt::all(), xt::all(), 0);
+  xt::xtensor<double, 3> dphi = xt::view(basis, xt::range(1, tdim + 1), xt::all(), xt::all(), 0);
```
Comment (jorgensd) [inline cpp/kernels_non_const_coefficient.hpp:58]:
```suggestion

```

File: cpp/kernels_non_const_coefficient.hpp [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+  // Create Finite elements for coefficient functions and tabulate shape functions
+  int num_coeffs = coeffs.size();
+  std::vector<int> offsets(num_coeffs + 1);
+  offsets[0] = 0;
+  for (int i = 1; i < num_coeffs + 1; i++)
+  {
+    offsets[i] = offsets[i - 1] + coeffs[i - 1]->function_space()->element()->space_dimension();
+  }
```
Comment (jorgensd) [inline cpp/kernels_non_const_coefficient.hpp:67]:
```suggestion
    offsets[i] = offsets[i - 1] + coeffs[i - 1]->function_space()->element()->space_dimension();
```

File: cpp/kernels_non_const_coefficient.hpp [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+
+    // Compute Jacobian, its inverse and the determinant
+    dolfinx_cuas::math::compute_jacobian(dphi0_c, coord, J);
+    double detJ = std::fabs(dolfinx_cuas::math::compute_determinant(J));
+
+    // Main loop
+    for (std::size_t q = 0; q < weights.size(); q++)
+    {
+      double w0 = 0;
+      for (int i = 0; i < num_coeffs; i++)
```
Comment (jorgensd) [inline cpp/kernels_non_const_coefficient.hpp:107]:
```suggestion
      //  For each coefficient (assumed scalar valued), compute  sum_{j=0}^{num_dofs} c^j phi^j(x_q)
      for (int i = 0; i < num_coeffs; i++)
```

File: cpp/kernels_non_const_coefficient.hpp [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+      {
+        for (int j = offsets[i]; j < offsets[i + 1]; j++)
+        {
+          w0 += c[j] * phi_coeffs(q, j);
+        }
```
Comment (jorgensd) [inline cpp/kernels_non_const_coefficient.hpp:112]:
```suggestion
          w0 += c[j] * phi_coeffs(q, j);
```

**Pull request merged**